src/contracts/deposits-registry/sanity-checker/sanity-checker.service.ts

```diff
@@ -0,0 +1,127 @@
+import { BlockchainCheckerService } from './blockchain-checker/blockchain-checker.service';
+import { DepositIntegrityCheckerService } from './integrity-checker';
+
+@Injectable()
+export class DepositRegistrySanityCheckerService {
+  constructor(
+    @Inject(WINSTON_MODULE_NEST_PROVIDER) private logger: LoggerService,
+    private blockchainSanityChecker: BlockchainCheckerService,
+    private depositsIntegrityChecker: DepositIntegrityCheckerService,
+  ) {}
+
+  public async initialize(initialEventsCache: VerifiedDepositEventsCache) {
+    await this.depositsIntegrityChecker.initialize(initialEventsCache);
+  }
+
+  private async indexEventsChunk(events: VerifiedDepositEvent[]) {
+    return await this.depositsIntegrityChecker.putFinalizedEvents(events);
+  }
+  // putLatestEvents
+  private async checkFreshEventsChunk(
+    blockNumber: number,
+    events: VerifiedDepositEvent[],
+  ) {
+    return await this.depositsIntegrityChecker.checkLatestRoot(
+      blockNumber,
+      events,
+    );
+  }
+
+  private findReorganization(
+    blockNumber: number,
+    blockHash: string,
+    events: VerifiedDepositEvent[],
+  ) {
+    const event = this.blockchainSanityChecker.findReorganizedEvent(
+      events,
+      blockNumber,
+      blockHash,
+    );
+
+    if (event) {
+      this.logger.error('Reorganization found in deposit event', {
+        blockHash: event.blockHash,
+        blockNumber: event.blockNumber,
+        depositDataRoot: event.depositDataRoot,
+      });
+      return true;
+    }
+    return false;
+  }
+
+  public verifyCacheBlock(
+    cachedEvents: VerifiedDepositEventsCache,
+    currentBlock: number,
+  ) {
+    const isCacheValid = this.blockchainSanityChecker.validateCacheBlock(
+      cachedEvents,
+      currentBlock,
+    );
+
+    const blocks = {
+      cachedStartBlock: cachedEvents.headers.startBlock,
+      cachedEndBlock: cachedEvents.headers.endBlock,
+      currentBlock,
+    };
+
+    if (isCacheValid) {
+      this.logger.log('Deposit events cache has valid age', blocks);
+    }
+
+    if (!isCacheValid) {
+      this.logger.error(
+        'Deposit events cache is newer than the current block',
+        blocks,
+      );
+    }
+
+    return isCacheValid;
+  }
+
+  public async verifyEventsChunk(
+    chunkStartBlock: number,
+    chunkToBlock: number,
+    events: VerifiedDepositEvent[],
+  ) {
+    if (!events.length) return;
+
+    const tree = await this.indexEventsChunk(events);
+
+    this.logger.log('Deposit events chunk was verified', {
+      chunkStartBlock,
+      chunkToBlock,
+      depositRoot: tree.getRoot(),
+    });
+  }
+
+  public async verifyFreshEvents(
+    blockNumber: number,
+    blockHash: string,
+    events: VerifiedDepositEvent[],
+  ) {
+    const isReorgFound = this.findReorganization(
+      blockNumber,
+      blockHash,
+      events,
+    );
+
+    if (isReorgFound) return false;
+
+    const isDepositRootMatches = await this.checkFreshEventsChunk(
```

Comment:
chunk in name gives impression that in checkFreshEventsChunk we verified only part of events